src/device_builder.rs

```diff
@@ -108,133 +158,17 @@ impl<'a, B: UsbBus> UsbDeviceBuilder<'a, B> {
-            self.config.manufacturer = None;
-            return self;
-        }
-
-        assert!(
-            manufacturer_ls.len() <= 16,
-            "Not support more than 16 \"manufacturer\"s"
-        );
-
-        // do list length check only if user already specify "extra_lang_ids"
-        if let Some(extra_lang_ids) = self.config.extra_lang_ids {
-            assert!(
-                manufacturer_ls.len() == extra_lang_ids.len() + 1,
-                "The length of \"product\" list should be one more than \"extra_lang_ids\" list",
-            )
-        }
-
-        self.config.manufacturer = Some(manufacturer_ls);
-
-        self
-    }
-
-    /// Sets the product name string descriptor.
-    ///
-    /// the first string should always be in English, the language of rest strings
-    /// should be pair with what inside [.extra_lang_ids()](Self::extra_lang_ids)
-    ///
-    /// Default: (none)
-    pub fn product(mut self, product_ls: &'a [&'a str]) -> Self {
-        if product_ls.len() == 0 {
-            self.config.product = None;
-            return self;
-        }
-
-        assert!(
-            product_ls.len() <= 16,
-            "Not support more than 16 \"product\"s"
-        );
-
-        // do list length check only if user already specify "extra_lang_ids"
-        if let Some(extra_lang_ids) = self.config.extra_lang_ids {
-            assert!(
-                product_ls.len() == extra_lang_ids.len() + 1,
-                "The length of \"product\" list should be one more than \"extra_lang_ids\" list",
-            )
-        }
-
-        self.config.product = Some(product_ls);
-
-        self
-    }
-
-    /// Sets the serial number string descriptor.
-    ///
-    /// the first string should always be in English, the language of rest strings
-    /// should be pair with what inside [.extra_lang_ids()](Self::extra_lang_ids)
-    ///
-    /// Default: (none)
-    pub fn serial_number(mut self, serial_number_ls: &'a [&'a str]) -> Self {
-        if serial_number_ls.len() == 0 {
-            self.config.serial_number = None;
-            return self;
-        }
-
-        assert!(
-            serial_number_ls.len() <= 16,
-            "Not support more than 16 \"serial_number\"s"
-        );
-
-        // do list length check only if user already specify "extra_lang_ids"
-        if let Some(extra_lang_ids) = self.config.extra_lang_ids {
-            assert!(
-                serial_number_ls.len() == extra_lang_ids.len() + 1,
-                "The length of \"serial_number\" list should be one more than \"extra_lang_ids\" list",
-            )
-        }
-
-        self.config.serial_number = Some(serial_number_ls);
-
-        self
+    /// # Note
+    /// Up to 16 languages may be provided.
+    pub fn strings(mut self, descriptors: &[StringDescriptors<'a>]) -> Result<Self, BuilderError> {
+        // The 16 language limit comes from the size of the buffer used to provide the list of
```

Comment:
Super-minor nit/question: does the size of that buffer ([this one](https://github.com/rust-embedded-community/usb-device/pull/130/files#diff-36327182a60f061fcbee1b4da122b99e82b6d4d213e4b3e94bf297354680a720R550)?) have documentation that explains this consequence?  Does the limit originate in the USB spec?

Reply:
I actually don't think language IDs are even _in_ the specification anymore, so there's not really anything limiting more than 16 languages. It's from USB 1.0 from what I can tell and has largely been deprecated (not sure in favor of what). Even finding the list of lang IDs is hard nowadays. It makes me wonder if we should even support multiple langs, but I guess it makes sense for legacy systems?